File: IntelPresentMon/AppWinUI/PresentMonUI.Core/AutomaticTargeting.cs
namespace PresentMon.UI.Core;

public sealed record GpuProcessSample(int Pid, double RunningTime);

public static class AutomaticTargeting
{
    public static readonly TimeSpan Interval = TimeSpan.FromSeconds(1);

    // The caller supplies a revision check to discard scans superseded by user input.
    public static async Task UpdateAsync(
        Func<Task<IReadOnlyList<GpuProcessSample>>> sample,
        Func<int[], Task<IReadOnlyList<int>>> probeGpuBusy,
        Func<bool> isCurrent,
        Func<int?, Task> select)
    {
        var samples = await sample();
        if (!isCurrent()) return;
        // Idle/invalid loads cannot win; avoid starting frame tracking for them.
        var active = samples.Where(s => double.IsFinite(s.RunningTime) && s.RunningTime > 0).ToArray();
        // An empty request also releases trackers left over from the previous scan.
        var measurable = (await probeGpuBusy(active.Select(s => s.Pid).Distinct().ToArray())).ToHashSet();
        if (!isCurrent()) return;
        var top = active.Where(s => measurable.Contains(s.Pid))
            .OrderByDescending(s => s.RunningTime).ThenBy(s => s.Pid).FirstOrDefault();
        // Eligibility controls switching; an empty result preserves the current target and capture.
        if (top is not null) await select(top.Pid);
    }
}

File: IntelPresentMon/AppWinUICpp/Core/AutomaticTargeting.cpp
// Copyright (C) 2026 Intel Corporation
// SPDX-License-Identifier: MIT
#include "AutomaticTargeting.h"

#include <algorithm>
#include <cmath>
#include <unordered_set>

namespace pmon::ui::core
{
    void AutomaticTargeting::Update(const Sample& sample, const ProbeGpuBusy& probeGpuBusy,
        const IsCurrent& isCurrent, const Select& select)
    {
        const auto samples = sample();
        if (!isCurrent()) {
            return;
        }

        std::vector<GpuProcessSample> active;
        std::vector<int> candidatePids;
        std::unordered_set<int> seen;
        for (const auto& entry : samples) {
            if (std::isfinite(entry.RunningTime) && entry.RunningTime > 0.) {
                active.push_back(entry);
                if (seen.insert(entry.Pid).second) {
                    candidatePids.push_back(entry.Pid);
                }
            }
        }

        const auto measurableList = probeGpuBusy(candidatePids);
        if (!isCurrent()) {
            return;
        }
        const std::unordered_set<int> measurable{ measurableList.begin(), measurableList.end() };
        const auto top = std::min_element(active.begin(), active.end(), [&measurable](const auto& left, const auto& right) {
            const auto leftEligible = measurable.contains(left.Pid);
            const auto rightEligible = measurable.contains(right.Pid);
            if (leftEligible != rightEligible) {
                return leftEligible;
            }
            if (!leftEligible) {
                return false;
            }
            if (left.RunningTime != right.RunningTime) {
                return left.RunningTime > right.RunningTime;
            }
            return left.Pid < right.Pid;
        });
        if (top != active.end() && measurable.contains(top->Pid)) {
            select(top->Pid);
        }
    }
}
